### PR TITLE
Account Settings

### LINKED ISSUE #370: The ability to just not show NSFW content
Maybe I'm missing it in settings, but it would be nice if you could choose to specifically not show NSFW content at all in Mlem, rather than just blurring it

### PR DESCRIPTION
**Pull request by Sjmarf**
Partially implemented Account Settings. 

- Partially addresses #770
- Closes #370

## New Options

The following settings changes are now possible within Mlem. It is only possible to edit these settings on instances running 0.18.0 and later - on older instances, there are some API differences that it would be a pain to account for, so I've decided to not support editing of account settings on those instances (most instances are running 0.18.0 these days anyways). A suitable message is shown if you try to edit settings on a <0.18.0 instance.

- Change display name, bio & Matrix ID
- Change email
- Change password
- Toggle "Show NSFW content" (#370)
- Toggle "Show Bot Accounts"
- Change discussion languages (side note: if I disable 'undetermined' and refresh Lemmy doesn't actually disable it, but I get the same weird behaviour on lemmy-ui, so maybe it's a bug on their end?)
- Toggle "send notifications to email"
- Toggle whether the account is a bot account

## Other Changes

The 'Delete current account' option has been removed from the account switcher and added to the account settings page.

## Future Plans

There are some settings notably missing from this PR that I'd like to add later. In lieu of submitting a gigantic PR, I've decided to submit the changes I've got so far before adding anything further.

In the immediate future, I'd like to add the option to change the profile picture and banner. Later I might implement 2FA setup; viewing and editing the lists of blocked communities / users and the ability to import / export account settings. At some point I may also look at sprucing up the account switcher a little - perhaps showing profile pictures there would be a nice touch.

## Screenshots

<table>
  <tr>
    <td> 
    <td> 
    <td> 
   </tr> 
   <tr>
      <td> 
      <td> 
      <td> 
  </tr>
  <tr>
      <td> 
      <td> 
      <td> 
  </tr>
</table>

### CODE REVIEW

File: Mlem/Views/Tabs/Settings/Components/Views/Account/AccountSettingsView.swift [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+                .disabled(settingsDisabled)
+                
+//                Section {
+//                    NavigationLink { EmptyView() } label: {
+//                        Label("Blocked Commuities", systemImage: "house.fill").labelStyle(SquircleLabelStyle(color: .gray))
+//                    }
+//                    NavigationLink { EmptyView() } label: {
+//                        Label("Blocked Users", systemImage: "person.fill").labelStyle(SquircleLabelStyle(color: .gray))
+//                    }
+//                }
```
Comment (EricBAndrews) [inline Mlem/Views/Tabs/Settings/Components/Views/Account/AccountSettingsView.swift:91]: Dead code or TODO?

File: Mlem/Views/Tabs/Settings/Components/Views/Account/AccountSettingsView.swift [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+                    }
+                    NavigationLink(.settings(.accountAdvanced)) {
+                        Label("Advanced", systemImage: "gearshape.2.fill")
+                        .labelStyle(SquircleLabelStyle(color: .gray))
+                    }
+                } footer: {
+                    if settingsDisabled {
+                        // swiftlint:disable:next line_length
+                        Text("We don't support editing of account settings on instances running versions older than 0.18.0 due to API differences on those versions. Your instance is running version \(String(describing: siteInformation.version ?? .zero)). Sorry!")
+                            .foregroundStyle(.red)
```
Comment (EricBAndrews) [inline Mlem/Views/Tabs/Settings/Components/Views/Account/AccountSettingsView.swift:78]: Could be more concise:

`Account settings are only available on instances running 0.18.0 or above`

File: Mlem/Views/Tabs/Settings/Components/Settings View.swift [hunk 3]
```diff
@@ -21,8 +25,57 @@ struct SettingsView: View {
...
                     Section {
+                        NavigationLink(.settings(.currentAccount)) {
+                            HStack(spacing: 20) {
+                                AvatarView(
+                                    url: appState.profileTabRemoteSymbolUrl,
+                                    type: .user,
+                                    avatarSize: 60,
+                                    iconResolution: .unrestricted
+                                )
+                                    .padding(.vertical, -8)
```
Comment (EricBAndrews) [inline Mlem/Views/Tabs/Settings/Components/Settings View.swift:36]: The spacing on this feels kinda off--what if we bump the avatar size down to 48 and ditch the negative padding?

Current:
Screenshot 2023-12-07 at 12 12 28 PM

Proposed:
Screenshot 2023-12-07 at 12 11 51 PM

**Review by EricBAndrews — CHANGES_REQUESTED** [review]
Everything looks like it works as intended.

A couple nits, but looks good overall!

File: Mlem/Views/Tabs/Settings/Components/Views/Account/AccountSettingsView.swift (hunk 1 above)
Comment (Sjmarf) [reply Mlem/Views/Tabs/Settings/Components/Views/Account/AccountSettingsView.swift:91]: TODO. The block lists will go here in future - I'm just commenting it out to save implementation time later.

File: Mlem/Views/Shared/Links/AvatarView.swift [hunk 4]
```diff
@@ -14,43 +19,75 @@ struct AvatarView: View {
...
     init(
-        community: CommunityModel,
+        url: URL?,
+        type: AvatarType,
         avatarSize: CGFloat,
+        blurAvatar: Bool = false,
+        clipAvatar: Bool = true,
         lineColor: Color? = nil,
-        iconResolution: Int? = nil
+        lineWidth: CGFloat = 1,
```
[the remark this answers is not in the document]
Comment (Sjmarf) [reply Mlem/Views/Shared/Links/AvatarView.swift:33]: Huh - I was under the impression that `clipAvatar` disabled the `clipShape`. It seems I was wrong. Was that the intended behaviour at one point? I thought we were trying to disable clipping from Beehaw avatars and such?

File: Mlem/Views/Tabs/Settings/Components/Settings View.swift (hunk 3 above)
Comment (EricBAndrews) [reply Mlem/Views/Tabs/Settings/Components/Settings View.swift:36]: Definitely agree that keeping things center-aligned is better. I still think the spacing is super tight--personally I think Apple dropped the ball on that one, but I'm content to leave as-is in accordance with our "follow Apple's lead" principle

Comment (Sjmarf) [reply Mlem/Views/Tabs/Settings/Components/Settings View.swift:36]: I don't mind reducing the current account icon size a little - the padding certainly is a little tight at the edges. I'd rather keep the alignment as is, though

Comment (Sjmarf) [reply Mlem/Views/Tabs/Settings/Components/Settings View.swift]: How's this?
IMG_9724

**Review by EricBAndrews — APPROVED** [review]
Looks great! Outstanding work as always

**Pull request merged**